client/server/render/index.js

```diff
@@ -320,7 +339,7 @@ function isServerSideRenderCompatible( context ) {
 	return Boolean(
 		context.section?.isomorphic &&
 			! context.user && // logged out only
-			context.layout
+			( context.layout || context.hasLayout ) // A layout was generated or we have one cached.
```

Comment:
Do we really need to add this condition here? Currently we never skip the assignment of `context.layout`, i.e., building the JSX element that should be rendered. Only the data fetching. Therefore the `context.layout` is always still there. We'll just won't use it when we already have cached markup. If the cached markup is ever evicted from the cache, we need `context.layout` to build the markup again.

Reply:
I found that we do need it, or the pipeline fails to use the SSR layout and the client renders the page -- at least in some cases. (In my testing, SSR stopped using the layout after two requests.)

I think one of the middlewares I started skipping is the one which adds the layout.

Reply:
Yes, this middleware is skipped in this PR: https://github.com/Automattic/wp-calypso/blob/264a0f3e1a2db2d7e556c76a00edcfb29b9a67f5/client/controller/shared.js#L28

I think this makes sense, as we don't need to do any of that work, but let me know what you think. Of course, we could also just store the cached layout inside of `layout`, and not overwrite it if it exists